# Exploratory Data Analysis with R 

Import relevant libraries

In [423]:
library(tidyverse)
library(data.table)
library(stringr)
library(readr)
library(plyr)
library(tidyr)
library(dplyr)

In [424]:
# reading in the data 
maxTemp_df = readRDS('../R/Data/Homog_monthly_max_temp_cleaned/5050920_FLIN FLON_MB.rds')
totalPrec_df = readRDS('../R/Data/Adj_monthly_total_prec_cleaned/5050920_FLIN FLON_MB.rds')
co2_alt_df = readRDS('../R/Data/Surface CO2 from NOAA_cleaned/co2_alt_surface-flask_1_ccgg_month.rds')


In [425]:
print(dim(maxTemp_df))
head(maxTemp_df, n=10)

[1] 91 18


Year Jan   Feb   Mar   Apr May  Jun  Jul  Aug  Sep  Oct  Nov   Dec   Annual
1  1927 NA    -13.5 NA    NA  NA   NA   NA   NA   NA   NA   NA    NA    NA    
2  1928 -8.9  NA    NA    NA  NA   21.4 27.8 NA   18.0 7.0  0.3   -7.9  NA    
3  1929 -21.9 -12.7 NA    7.1 13.1 20.9 24.6 24.6 14.5 11.0 -5.1  -15.2 NA    
4  1930 -21.4 -12.1 -2.6  8.2 13.8 21.3 25.9 25.5 12.8 2.9  -4.3  -5.8  5.4   
5  1931 -12.5 -5.6  NA    7.6 14.3 20.0 23.7 21.9 14.2 9.6  -3.6  -9.0  NA    
6  1932 -15.2 -15.0 -11.3 5.4 17.2 21.5 23.8 23.4 15.6 1.5  -10.5 -17.1 3.3   
7  1933 -18.6 -17.2 -6.6  3.2 15.4 21.1 22.2 22.5 14.0 2.4  -10.3 -25.5 1.9   
8  1934 -16.1 -11.5 NA    6.1 15.1 18.7 21.3 19.1 11.0 7.2  -4.6  -15.1 NA    
9  1935 -22.7 -5.7  -8.1  4.7 16.2 18.2 25.1 20.1 13.4 3.8  -11.2 -13.0 3.4   
10 1936 -23.9 -22.0 -5.3  1.9 14.3 18.8 25.4 21.7 11.9 2.9  -4.0  -16.5 2.1   
   Winter Spring Summer Autumn
1  NA     NA     NA     NA    
2  NA     NA     NA     8.4   
3  -14.2  NA     23.4   6.8   
4  -16.2  6.5    24.2   3.8   
5  -8.0   NA     21.9   6.7   
6  -13.1  3.8    22.9   2.2   
7  -17.6  4.0    21.9   2.0   
8  -17.7  NA     19.7   4.5   
9  -14.5  4.3    21.1   2.0   
10 -19.6  3.6    22.0   3.6

In [429]:
print(dim(co2_alt_df))
head(co2_alt_df, n=15)

[1] 386   4


site year month value 
1  ALT  1985 6     349.94
2  ALT  1985 7     343.98
3  ALT  1985 8     337.96
4  ALT  1985 9     339.09
5  ALT  1985 10    342.85
6  ALT  1985 11    345.66
7  ALT  1985 12    348.97
8  ALT  1986 1     351.34
9  ALT  1986 2     352.13
10 ALT  1986 3     352.74
11 ALT  1986 4     353.12
12 ALT  1986 5     353.32
13 ALT  1986 6     351.45
14 ALT  1986 7     345.76
15 ALT  1986 8     340.40

In [426]:
print(dim(totalPrec_df))
head(totalPrec_df, n=10)

[1] 91 18


Year Jan Feb Mar Apr  May  Jun   Jul   Aug   Sep   Oct  Nov  Dec Annual
1  1927 NA  NA  NA  36.5 15.7 13.7  103.0 65.5  86.3  16.1 NA   NA  NA    
2  1928 NA  NA  NA  13.3 33.3 28.3  55.8  35.0  13.9  14.2 2.6  NA  NA    
3  1929 NA  NA  NA  13.7 12.2 23.6  81.8  28.4  62.1  1.5  22.1 NA  NA    
4  1930 NA  NA  NA  19.4 61.2 100.7 57.3  42.8  88.8  NA   NA   NA  NA    
5  1931 NA  NA  NA  NA   40.8 101.1 57.1  69.6  92.7  46.2 NA   NA  NA    
6  1932 NA  NA  NA  NA   45.9 71.3  66.7  146.6 4.4   NA   NA   NA  NA    
7  1933 NA  NA  NA  NA   27.1 80.3  96.6  27.0  52.3  NA   NA   NA  NA    
8  1934 NA  NA  NA  25.6 38.0 123.4 69.1  19.0  20.5  25.2 NA   NA  NA    
9  1935 NA  NA  NA  13.1 51.9 55.0  52.3  37.0  47.9  17.8 NA   NA  NA    
10 1936 NA  NA  NA  2.4  38.2 54.4  91.7  97.5  112.3 7.6  NA   NA  NA    
   Winter Spring Summer Autumn
1  NA     NA     182.2  NA    
2  NA     NA     119.2  30.8  
3  NA     NA     133.7  85.8  
4  NA     NA     200.8  NA    
5  NA     NA     227.7  NA    
6  NA     NA     284.6  NA    
7  NA     NA     204.0  NA    
8  NA     NA     211.5  NA    
9  NA     NA     144.4  NA    
10 NA     NA     243.6  NA

## Constructing desired Dataframe 


To-DO - remove NAs for machine learning. mlr already does it...but ml wont 

In [419]:
# Purpose: Load data from cleaning step
load_cleaned_data <- function(year_to_start = 1960, month = 'Jan', var_name, file_name, data){
  
    nom_city <- ns[2]
    nom_prov <- unlist(strsplit(ns[3],'.rds'))
    ns = matrix(unlist(strsplit(file_name,'_',)),ncol = 3,byrow = TRUE) #name_split 
    var_names = paste(var_name, collapse = '_')
    
    # build input data frame. 
    input_df <- data.frame()
    
    df_1 <- data[[1]] # 1st dataframe includes target ... 

    years_greater<-df_1[as.numeric(as.character(df_1$Year))>=year_to_start,]
    y_temp <- years_greater %>% select(month) %>% unlist() %>% as.character() %>% as.numeric()
#   y_temp <- years_greater %>% select(month) %>% mutate_all(function(x) as.numeric(as.character(x)))
    x_1 <- years_greater %>% select('Year') %>% unlist()%>% as.character() %>% as.numeric()  #year always needed to join..could be independent from analysis
    
    temp_df <- data.frame('y_temp' = y_temp, 'x_year' = x_1)

    print(length(data))
    if(length(data) > 1){
        for (i in 2:length(data)){
            feature <- data[[i]] %>% select('Year', month) %>% mutate_all(function(x) as.numeric(as.character(x)))
            temp_df <- left_join(temp_df, feature, by = c('x_year' = 'Year'))
            names(temp_df)[i+1] <- paste0('x_',var_name[i+1])
            print(var_name[i+1])                                                              
        }
        
    }
    
    temp_df <- data.frame(temp_df, "city" = nom_city, "prov" = nom_prov, 'var_name' = var_names)
                                                            
#     input_df <- rbind(input_df, temp_df)
    
    return(temp_df)
}
                                                                      

In [420]:
file_name <- "5050920_FLIN FLON_MB"
meas <- c("minTemp", 'year','totalPrec')
# meas <- c("minTemp", 'year')
year_to_start <- 1960
month <- 'Jan'
data <- list(maxTemp_df,totalPrec_df)
# data <- list(maxTemp_df)
# data[[2]]
input_df <- load_cleaned_data(year_to_start,month, meas,file_name, data)
head(input_df)

[1] 2


Warning message in (function (x) :
"NAs introduced by coercion"

[1] "totalPrec"


y_temp x_year x_totalPrec city      prov var_name              
1 -17.1  1960    7.9        FLIN FLON MB   minTemp_year_totalPrec
2 -15.5  1961    4.4        FLIN FLON MB   minTemp_year_totalPrec
3 -19.1  1962   22.1        FLIN FLON MB   minTemp_year_totalPrec
4 -21.5  1963   25.7        FLIN FLON MB   minTemp_year_totalPrec
5 -14.2  1964   43.2        FLIN FLON MB   minTemp_year_totalPrec
6 -19.7  1965   17.6        FLIN FLON MB   minTemp_year_totalPrec

In [421]:
regression <- function(input_df){
  city_prov_vector <- unique(input_df[,c("city", 'prov')])
  city_vector <- city_prov_vector[, 'city']
  prov_vector <- city_prov_vector[, 'prov']
  var_name <- unique(input_df$var_name)
    
  print(feature <- input_df%>% select(starts_with('x')) %>% names() )
    print(length(feature))
  output_df <- data.frame()

  

  for (i in 1:nrow(city_prov_vector)){
    index <- which(input_df$prov==prov_vector[i]
                   & input_df$city == city_vector[i])
      
      if(length(feature) == 1)
        fit <- lm(y_temp[index]~x_year[index], data = input_df)
        
      else if(length(feature) == 2)
        fit <- lm(y_temp[index]~input_df[,feature[1]][index]+input_df[,feature[2]][index], data = input_df)

    b <- data.frame("intercept" = fit$coefficients[1], "slope" = fit$coefficients[2])
    R_2 <- data.frame("r.squared" = as.numeric(unlist(summary(fit)$r.squared)))
    # CIs <- ci(fit, 0.95, alpha=1-0.95, na.rm = TRUE)
    critical_value <- qt((1-0.95)/2, (nrow(fit$model)-1))
    standard_error <- summary(fit)$coef[,2][2]
    margin_error <- critical_value*standard_error
    estimate <- summary(fit)$coef[,1][2]
    CI_lower <-  estimate + margin_error
    CI_upper <- estimate - margin_error
    variance <- (standard_error)^2
    curr_results_df <- data.frame("city"=city_vector[i],'prov' = prov_vector[i],  
                                  b,"r.squared"=R_2,CI_lower, CI_upper,variance,
                                  "n"=nrow(fit$model), 'var_name' = var_name,  row.names = NULL)
    output_df <- rbind(output_df,curr_results_df)
  }
  return(output_df)
}

In [422]:
output_df <- regression(input_df)
output_df

[1] "x_year"      "x_totalPrec"
[1] 2
[1] "im here 2"


city      prov intercept slope     r.squared CI_lower   CI_upper 
1 FLIN FLON MB   -230.6669 0.1082407 0.2048238 0.04992736 0.1665539
  variance     n  var_name              
1 0.0008473617 57 minTemp_year_totalPrec

    city	prov	intercept	slope	r.squared	CI_lower	CI_upper	variance	    n	var_name
FLIN FLON	MB	-225.9276	0.1056723	0.2013154	0.04934987	0.1619947	0.0007911016	58	minTemp_year


    city	prov	intercept	slope	r.squared	CI_lower	CI_upper	variance	     n	var_name
FLIN FLON	MB	-230.6669	0.1082407	0.2048238	0.04992736	0.1665539	0.0008473617	57	minTemp_year_totalPrec



## Saving files under new names and new and directories
Extracting the station number, city and province for file name use. 

In [31]:
df = read.delim('R/Data/Homog_monthly_max_temp/mx230N002.txt', skip = 0, header = FALSE, as.is=TRUE, dec=".", sep = ",", na.strings=c(" ", "",'NA'), strip.white = TRUE)

(station_number <- select(df, V1)[1,1])
(city <- select(df, V2)[1,1])
(province <- select(df, V3)[1,1])


[1] "230N002"

[1] "LUPIN"

[1] "NU"

In [32]:
stationNum_city_prov <- paste(select(df, V1)[1,1], trimws(select(df, V2)[1,1]), province <- select(df, V3)[1,1], sep='_')
stationNum_city_prov

[1] "230N002_LUPIN_NU"

## Result 

338 text files were cleaned, per directory
* Directories: Homog_monthly_min_temp, Homog_monthly_max_temp, Homog_monthly_mean_temp

As shown above, new text files and directories are created
* New Directories: Homog_monthly_min_temp_cleaned, Homog_monthly_max_temp_cleaned, Homog_monthly_mean_temp_cleaned

 

## Intended use 
All data is cleaned prior to the shinyApp being used. The cleaned data is reloaded and processed into a single data frame on app load. 


### Input data frame. 
Depending on the month and year the user selects, the data is trimmed before any statistical method.


![input df](input_df.PNG)

### Output data frame. 
After the desired statistical analysis, we shift our focus to different variables and a new data frame.


![output df](output_df.PNG)

## In conclusion, the data does not appear to have many challenges. The ease and use of the tidy verse package, is one that can be consistently used. Helping many, understand the data they are dealing with.